node/src/components/small_network/message.rs

```diff
@@ -1,17 +1,32 @@
-use std::fmt::{self, Debug, Display, Formatter};
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    net::SocketAddr,
+};
 
 use serde::{Deserialize, Serialize};
 
 #[derive(Clone, Debug, Deserialize, Serialize)]
 pub enum Message<P> {
-    Handshake { network_name: String },
+    Handshake {
+        /// Network we are connected to.
+        network_name: String,
+        /// The public address of the node connecting.
+        public_address: SocketAddr,
```

Comment:
I don't understand why we need this information sent to us – we know from what IP and port combination this message comes from so why do we need to send it as well?

Reply:
A an outgoing TCP connection is made from a random port. Let's assume we have two nodes `A` with IP `2.2.2.2` and `B` with IP `7.7.7.7`, both listening on a fictional node-port `9999`.

Listening addresses are thus `2.2.2.2:9999` and `7.7.7.7:9999`.

If `A` connects to be, the OS of A will allocate a random port for the outgoing connection, e.g. `40318`. So the connection will be

```
[A] 2.2.2.2:40318   ----->  7.7.7.7:9999 [B]
```

Thus the peer address of the incoming connection will show up as `2.2.2.2:40318`. However, Node B needs to connect back to `2.2.2.2:9999` instead, so we need to send the address.

When `B` connects back, it will also be from a random port, resulting in something like

```
[A] 2.2.2.2:40318   ----->  7.7.7.7:9999 [B]
[A] 2.2.2.2:9999   <-----   7.7.7.7:50223 [B]
```